_posts/2020-08-05-hive-acid-original-files.md

```diff
@@ -0,0 +1,182 @@
+|n_nationkey : int, |struct {            |struct {                      |
+|n_name : string,   |n_nationkey : int,  |operation : int,              |
+|n_regionkey : int, |n_name : string,    |originalTransaction : bigint, |
+|n_comment : string` |n_regionkey : int, |bucket : int,                 |
+|                   |n_comment : string  |rowId : bigint,               |
+|                   |}                   |currentTransaction : bigint,  |
+|                   |                    |row : struct {                |
+|                   |                    |    n_nationkey : int,        |
+|                   |                    |    n_name : string,          |
+|                   |                    |    n_regionkey : int,        |
+|                   |                    |    n_comment : string        |
+|                   |                    |  }                           |
+|                   |                    |}                             |
+|                   |                    |                              |
+
+
+
+Before Presto 340, Presto used to fail the query if it reads from a Hive
+transactional table having original files.
+
+# Update and delete support on original files
+
+Hive achieves updates/deletes on a row in original files by synthetically
+generating ACID columns for original files. Presto follows the same mechanism of
+generating ACID columns synthetically as discussed later.
+
+## ACID column generation on original files
+
+Files in Hive ACID format have 5 ACID columns, but we need only 3 columns i.e.
+`originalTransactionId`, `bucketId` and `rowId` to uniquely identify a row. In
+this section, we will see how these 3 columns are synthetically generated for
+original files.
+
+### Original transaction ID
+
+An original transaction ID is the write ID when a record is first created. For
+original files, the original transaction ID is always 0.
+
+### Bucket ID
+
+Bucket ID is retrieved from the original file name. For the original file
+`0000ABC_DEF` or `0000ABC_DEF_copy_G`, the bucket ID will be `ABC`.
+
+### Row ID
+
+To calculate the row ID, the total row count of all the original files, which
+come before the current one in lexicographical order, is calculated.
+Then, the row ID is equal to the sum of the value calculated and local row ID in
+the current original file.
+
+Here is an example to calculate the global Row ID of the 3rd row of an original
+File `000000_0_copy_2`.
+~~~~
+000000_0            -> 	X1 Rows (returned by ORC footer field numberOfRows)
+
+000000_0_copy_1     -> 	X2 Rows (returned by ORC footer field numberOfRows)
+
+000000_0_copy_2     ->	[ Row 0 ]
+                        [ Row 1 ]
+                        [ Row 2 ]   <- Local Row ID (returned filePosition in OrcRecordReader) = 2, Global Row ID = (X1+X2+2)
+                        [ Row 3 ]
+
+000000_0_copy_3     ->  X4 Rows
+~~~~
+
+>**Note:** As we see, additional computations are required to generate row IDs
```

Comment:
Does this render as desired?

Reply:
![image](https://user-images.githubusercontent.com/4463893/93222445-9675b500-f78c-11ea-8e0b-b9c2fe822a4f.png)

yes, it is getting rendered as desire, attached the picture.